sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java

```diff
@@ -117,60 +127,31 @@ public void verifyUnmodified() {
     }
 
     private void verifyUnmodifiedThrowingCheckedExceptions() throws CoderException {
-      // If either object believes they are equal, we trust that and short-circuit deeper checks.
-      if (Objects.equals(possiblyModifiedObject, clonedOriginalObject)
-          || Objects.equals(clonedOriginalObject, possiblyModifiedObject)) {
-        return;
-      }
-
-      // Since retainedObject is in general an instance of a subclass of T, when it is cloned to
-      // clonedObject using a Coder<T>, the two will generally be equivalent viewed as a T, but in
-      // general neither retainedObject.equals(clonedObject) nor clonedObject.equals(retainedObject)
-      // will hold.
-      //
-      // For example, CoderUtils.clone(IterableCoder<Integer>, IterableSubclass<Integer>) will
-      // produce an ArrayList<Integer> with the same contents as the IterableSubclass, but the
-      // latter will quite reasonably not consider itself equivalent to an ArrayList (and vice
-      // versa).
-      //
-      // To enable a reasonable comparison, we clone retainedObject again here, converting it to
-      // the same sort of T that the Coder<T> output when it created clonedObject.
-      T clonedPossiblyModifiedObject = CoderUtils.clone(coder, possiblyModifiedObject);
-
-      // If deepEquals() then we trust the equals implementation.
-      // This deliberately allows fields to escape this check.
-      if (Objects.deepEquals(clonedPossiblyModifiedObject, clonedOriginalObject)) {
-        return;
+      // Since there is not guarantee that cloning an object via the coder will
+      // return the exact same type as value, We are cloning the possiblyModifiedObject
+      // before getting it's structural value. This way we are guarantee to compare the same
+      // types.
+      T possiblyModifiedClonedValue = CoderUtils.clone(coder, possiblyModifiedObject);
+      Object newStructuralValue = coder.structuralValue(possiblyModifiedClonedValue);
+      if (originalStructuralValue.equals(newStructuralValue)) {
+          return;
       }
-
-      // If not deepEquals(), the class may just have a poor equals() implementation.
-      // So we next try checking their serialized forms. We re-serialize instead of checking
-      // encodedObject, because the Coder may treat it differently.
-      //
-      // For example, an unbounded Iterable will be encoded in an unbounded way, but decoded into an
-      // ArrayList, which will then be re-encoded in a bounded format. So we really do need to
-      // encode-decode-encode retainedObject.
-      if (Arrays.equals(
-          CoderUtils.encodeToByteArray(coder, clonedOriginalObject),
-          CoderUtils.encodeToByteArray(coder, clonedPossiblyModifiedObject))) {
-        return;
+      if (possiblyModifiedClonedValue.equals(clonedOriginalValue)){
```

Comment:
My understanding was that this would collapse all the way, to:

```T possiblyModifiedClonedValue = ...
Object possiblyModifiedStructuralValue = ...
if (!originalStructuralValue.equals(possiblyModifiedStructuralValue)) {
  illegalMutation(...)
}
```

thus only utilizing the `Coder's` understanding of equality.